app/models/manageiq/providers/openstack/storage_manager/cinder_manager/event_target_parser.rb

```diff
@@ -0,0 +1,44 @@
+class ManageIQ::Providers::Openstack::StorageManager::CinderManager::EventTargetParser
+  attr_reader :ems_event
+
+  # @param ems_event [EmsEvent] EmsEvent object to be parsed to derive an object to be refreshed
+  def initialize(ems_event)
+    @ems_event = ems_event
+  end
+
+  # Parses all targets present in the EmsEvent givin in the initializer
+  # @return [Array] Array of ManagerRefresh::Target objects
+  def parse
+    parse_ems_event_targets(ems_event)
+  end
+
+  private
+
+  # Parses list of ManagerRefresh::Target(s) out of the given EmsEvent
+  #
+  # @param ems_event [EmsEvent] EmsEvent object
+  # @return [Array] Array of ManagerRefresh::Target objects
+  def parse_ems_event_targets(ems_event)
+    target_collection = ManagerRefresh::TargetCollection.new(:manager => ems_event.ext_management_system.parent_manager, :event => ems_event)
+
+    # there's almost always a tenant id regardless of event type
+    collect_identity_tenant_references!(target_collection, ems_event)
+
+    # WIP
```

Comment:
right, this will be for standalone block storage based entities, can we also try to extract targets as part of Vm in collector?

e.g. amazon look both in db and API:
https://github.com/Ladas/manageiq-providers-amazon/blob/52636410835d6319627afd5e489cefe199e81a01/app/models/manageiq/providers/amazon/inventory/collector/target_collection.rb#L203

https://github.com/Ladas/manageiq-providers-amazon/blob/52636410835d6319627afd5e489cefe199e81a01/app/models/manageiq/providers/amazon/inventory/collector/target_collection.rb#L249

# And seems like I miss extracting it from the db
---

It's currently the only way how to get full targeted refresh of a Vm

Reply:
Yeah absolutely, volumes should be pulled as part of that.